core/headerchain.go

```diff
@@ -196,6 +187,18 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
 	if err := WriteHeader(hc.chainDb, header); err != nil {
 		glog.Fatalf("failed to write header contents: %v", err)
 	}
```

Comment:
The loop above still rewrites canonical hashes before the header is written. A safer change would be to move the writes of Header and TD up so they happen before the canonical numbers are dealt with.